apis/Google.Cloud.Firestore.Data/docs/datamodel.md

```diff
@@ -0,0 +1,173 @@
+Similarly, you can serialize as an anonymous type then deserialize as an attributed class.
+
+## Mapping with attributed classes
+
+If you apply the `[FirestoreData]` attribute to a class, the Firestore library for .NET can use it for serialization (when sending data to the server)
+and deserialization (when retrieving data). All public instance properties with the `[FirestoreProperty]` attribute applied are serialized.
+
+For example, to model a city as a document, you might have a class like this:
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.DataModel.txt#AttributedClass)]
+
+Note how the `IsCapital` property specifies the name in the attribute; this is the field name that will be in the stored document.
+This allows you to use idiomatic names in your .NET code, but match whatever field name is used in your Firestore database.
+
+You might then use it to create a document and then fetch it like this:
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.DataModel.txt#AttributedClassUsage)]
+
+In order to deserialize a map as an attributed class, the class must have a public parameterless constructor. (The C# compiler provides
+one by default if no other constructors are specified.) The properties in the map must each have an attributed, public instance property with a public setter.
+
+## Mapping with dictionaries
+
+A map can be represented as a `Dictionary<string, object>` - or if you're only storing values of a particular type, a `Dictionary<string, int>`, `Dictionary<string, string>` etc.
+The key type for the dictionary must always be `string`, as the Firestore field name is used as the key.
+
+Dictionaries can be passed to the various document creation and modification methods to represent the data, and `DocumentSnapshot.ToDictionary` deserializes
+document data to a dictionary representation.
+
+The equivalent city code using dictionaries would look like this:
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.DataModel.txt#DictionaryUsage)]
+
+## Mapping with anonymous types
+
+Anonymous types can be used for serialization, but not deserialization. They are particularly useful to specify partial updates, or
+to populate data which isn't then read within the same codebase.
+
+[!code-cs[](obj/snippets/Google.Cloud.Firestore.Data.DataModel.txt#AnonymousTypeUsage)]
+
+# Sentinel values
+
+So far all the values we've looked at have been known by the C# code. There are two *sentinel values* available which behave slightly differently.
+
+## Server-side timestamp
+
+When you update a document using the server-side timestamp sentinel value, the actual timestamp that's recorded is the commit time
+according to the Firestore server.
+
+For attributed classes, you can specify `SentinelValue = SentinelValue.ServerTimestamp` in the attribute declaration. Usually
+this will be on a property of type `Timestamp` (or `DateTime` or `DateTimeOffset`) so that you can retrieve the timestamp later.
```

Comment:
What happens if the type is something else. `String` or `double` say?

Reply:
Same as any other property - I've clarified that behaviour earlier.